db/memorystore.go

```diff
@@ -116,6 +119,31 @@ func (m *MemoryStore) AddAccount(acct *core.Account) (int, error) {
 	return len(m.accountsByID), nil
 }
 
+func (m *MemoryStore) ChangeAccountKey(acct *core.Account, newKey *jose.JSONWebKey) (error, *core.Account) {
+	m.Lock()
+	defer m.Unlock()
+
+	oldKeyID, err := keyToID(acct.Key)
+	if err != nil {
+		return err, nil
+	}
+
+	newKeyID, err := keyToID(newKey)
+	if err != nil {
+		return err, nil
+	}
+
+	if otherAccount, present := m.accountsByKeyID[newKeyID]; present {
+		return fmt.Errorf("New public key is already in use"), otherAccount
```

Comment:
In general it's not good Go practice to return both an error and an instance of the non-error return type. 

It would be better here to return just an `error` from this function. Return only `nil` when everything goes OK and return `existingAccountError{otherAccount.ID}` for a key conflict. You would then define your own `existingAccountError` type that implements the error interface, holds an account ID, and prints a good message about there being an existing account with the given ID.

In the calling code you can specifically check whether `err` is of type `existingAccountError`, and if it is, cast it to that type and extract the existing account ID to return as a `Location` header.

It's a bit more work but I think it will be a cleaner design overall

Reply:
I guess it must be `ExistingAccountError`, otherwise I cannot use it both in `db` and `wfe`.

Reply:
@felixfontein Yup! You're correct. It should be an exported type for that reason. Sorry I missed that :-)

Reply:
@cpu Great! I'm learning a bit more of Go every time I work on this ;)